### DISCUSSION TITLE
Can we use the extra link feature in Airflow to download a specific file from S3 for the user

### DISCUSSION THREAD

**Discussion by piyush-daga** [apache/airflow#18646] [Q&A] [answered]
Hi, some context around the problem statement.

Most of our tasks are running `BashOperator` and upon finishing they tend to make `on_success_callback` to send specific metadata to S3. That S3 bucket is a private one and not a lot of analysts in our org have AWS accounts setup to access the same.

So, we were looking at providing the download functionality within Airflow (as it is behind Okta auth with Airflow role management). We saw `extra links` feature which adds a link to the task instance context menu which would be a perfect place for this metadata download, as it would also have access to the `task_instance` variable to make the correct s3 url.

But the functionally seems pretty limited with sparse documentation around it.

Is there some way we can leverage the task instance context menu or Airflow for that matter to provide one-click download buttons (like how the `Download Log (by attempts)` button exists).

Thanks any help is appreciated. I would also be happy to contribute back to the docs.

**Comment by potiuk** [comment, marked as answer]
I think the documentation is pretty good - including screenshots:
https://airflow.apache.org/docs/apache-airflow/stable/howto/define_extra_link.html

And here you can see an overview of those operators that implement extra links so that you can see some examples and inspiration https://airflow.apache.org/docs/apache-airflow-providers/core-extensions/extra-links.html

**Reply by piyush-daga** [reply]
The documentation is easy to find and I was easily able to search for `extra-links and examples` of the same.

My concern is that extra-links, can only be used to provide a link. Or rather the final output of the button should be a link.
I was thinking more along the lines of how we would do this in flask:

```python
return send_file('<file-loc>', as_attachment=True)
```

Is this something that can be done / exposed via `extra-links`? I am not able to find proper documentation around the same.

**Reply by potiuk** [reply]
The link is automatically rendered in the Airflow UI as a button. This is part of Airflow UI. Why do you want to the fileloc ? What's s the purpose of it ? Can you explain why link renderd in the UI would not work ?

**Reply by potiuk** [reply]
The flask/as attachment only makes sens if you serve files from flask server. And when you provide the links it it entirely up to the sever to provide content disposition.  It's really nothing Airflow can do if the log files are served by the external system

**Reply by potiuk** [reply]
Ok. I see from the other theread that you would like to use the log to download it FROM airflow not external system . But this is a bad idea. It is much better to generate the extra link to download the file from S3. Even if it private bucket you can easilyy generate temporary signed link in 'extra link' so that it can be used to download the file. This is typically how it is done

**Reply by piyush-daga** [reply]
So, I'll go back to the basics of the problem statement. Th idea is that users accessing the Airflow UI are authenticated and we can share files from there without users having to access S3.

The solution I had in mind was more along these lines:
`Preferred:` The below function is available as a button for `BashOperator` using `extra-link`. 


```python
def download_button(self, conn, s3_hook, operator):
  # We get the id and email of the user trying to download the file some
  # And save the access / audit logs in the metastore (different schema)
  conn.execute_query("to save the id, email, file and time of access")

  # Construct the location of the file based on task-instance
  s3_loc = f'{airflow_folder}/{year}/{month}/{day}/{task-name}/...some other details'

  # Download the file from S3 and save in the local storage of the webserver pod
  s3_hook.download_file(bucket, s3_loc, '/tmp/..somewhere')

  # Share the file with the user
  return self.send_file('<file-loc>', as_attachment=True)

  # Find a way to delete the downloaded file after sharing.
```

The syntax does not conform exactly with extra-link's and is more pseudo-code

**Reply by piyush-daga** [reply]
> Even if it private bucket you can easilyy generate temporary signed link in 'extra link' so that it can be used to download the file. This is typically how it is done

This seems like a very good idea, I will have to run it past the security team once, but looks doable. Thanks for the help, and I apologise for the confusion caused in explaining the initial problem statement

**Comment by potiuk** [comment]
You can add such extra link as either a Plugin: https://airflow.apache.org/docs/apache-airflow/stable/plugins.html?highlight=plugins or if you plan to release a pacakge of new hooks, operators etc. via providers https://airflow.apache.org/docs/apache-airflow-providers/index.html (including detailed information and explanation on how to create custom providers).

**Comment by potiuk** [comment]
BTW. Why do you think the documentation is "sparse" ?  I found it immediately when I searched for it (I just typed "extra links" in seach of the docs). I would like to understand better, where the "sparse" documentation feeling comes frome @piyush-daga ? Especially that you are asking questions that seem to be already answered in the documentation (specifically the quesstion "Is there some way we can leverage the task instance context menu or Airflow for that matter to provide one-click download buttons (like how the Download Log (by attempts) button exists)." seems to be directly answered by the example there and screenshot - showing exactly this kind of on-click-download button.  But maybe we miss something obvious? 

So few questions:

* Was it difficult to find? 
* Did you try just google search or did you also try to directly look into our documentation (and if you tried, I would like to understand how and what was difficult there?
* Did you actually see that page(including the screenshot) which explains what extra links are before? 
* If you did  - was it not enough to answer your questions? What was missing there if not ?
* How do you think it can be corrected if you think what we have there is not enough ?

**Reply by piyush-daga** [reply]
Hi, @potiuk , thanks for taking the time for my question.
```
Is there some way we can leverage the task instance context menu or Airflow for that matter to provide one-click download buttons (like how the Download Log (by attempts) button exists).
```
Note, my intention is not to direct users to external systems, but rather to use the `extra-links` feature as a `download button`. At this point I'm not sure if the same is possible (with extra-links), maybe it is possible with `plugins`, but my knowledge around it is limited.

I will go through the `Download Log (by attempts)` code, and see if I find anything fruitful there.

With documentation being sparse I meant, every instance of `extra-link` I found, was using it for a single purpose, i.e. to construct a link and direct users to some external system (which is the advertised purpose). The docs do not provide enough information like if the flask object is available inside the function to (for example) send files to the user.
Also, the plugin documentation reads more like code documentation, than I how-to-guide for plugins, the same has been very difficult to find (even blog posts around it don't delve deep).
So I read through the plugin documentation, and did not have an intuitive idea of what all I can do with it, and what the code displayed there should do.

Maybe I'm missing something simple, I will definitely take time to read through the code and documentation more.
Please excuse my ignorance on this topic, I've found the Airflow docs to be most-helpful and updated.
